src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs

```diff
@@ -8847,5 +8851,214 @@ public static void UnsafeAwaitAwaiter<TAwaiter>(TAwaiter awaiter) where TAwaiter
+                using System.Threading.Tasks;
+
+                System.Console.WriteLine(await C.Handler());
+
+                class C
+                {
+                    public static Task<int> Handler()
+                    {
+                        try
+                        {
+                            return Throw(42);
+                        }
+                        catch (IntegerException ex)
+                        {
+                            return Task.FromResult(ex.Value);
+                        }
+                    }
+
+                #pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
+                    public static async Task<int> Throw(int value) => throw new IntegerException(value);
+                }
+
+                public class IntegerException(int value) : Exception(value.ToString())
+                {
+                    public int Value => value;
+                }
+                """;
+
+            var comp = CreateRuntimeAsyncCompilation(code);
+            var verifier = CompileAndVerify(comp, expectedOutput: ExpectedOutput("42", isRuntimeAsync: true), verify: Verification.Fails with
+            {
+                ILVerifyMessage = $$"""
+                    {{ReturnValueMissing("<Main>$", "0xf")}}
+                    """
+            });
+            verifier.VerifyIL("C.Handler()", """
+                {
+                  // Code size       25 (0x19)
+                  .maxstack  1
+                  .locals init (System.Threading.Tasks.Task<int> V_0)
+                  .try
+                  {
+                    IL_0000:  ldc.i4.s   42
+                    IL_0002:  call       "System.Threading.Tasks.Task<int> C.Throw(int)"
+                    IL_0007:  stloc.0
+                    IL_0008:  leave.s    IL_0017
+                  }
+                  catch IntegerException
+                  {
+                    IL_000a:  callvirt   "int IntegerException.Value.get"
+                    IL_000f:  call       "System.Threading.Tasks.Task<int> System.Threading.Tasks.Task.FromResult<int>(int)"
+                    IL_0014:  stloc.0
+                    IL_0015:  leave.s    IL_0017
+                  }
+                  IL_0017:  ldloc.0
+                  IL_0018:  ret
+                }
+                """);
+
+            comp = CreateRuntimeAsyncCompilation(code, options: TestOptions.DebugExe.WithDebugPlusMode(true));
+            verifier = CompileAndVerify(comp, expectedOutput: ExpectedOutput("42", isRuntimeAsync: true), verify: Verification.Fails with
+            {
+                ILVerifyMessage = $$"""
+                    {{ReturnValueMissing("<Main>$", "0x12")}}
+                    """
+            });
+            verifier.VerifyIL("C.Handler()", """
+                {
+                  // Code size       30 (0x1e)
+                  .maxstack  1
+                  .locals init (System.Threading.Tasks.Task<int> V_0,
```

Comment:
Would probably be good to have a ValueTask test as well as some custom task-like test. The return temps in those cases also need to be unwrapped, right?

Reply:
Only ValueTask<T>. Custom task-likes will go through the old codepath and get the full state machine.

If you don't mind, I'll add another test in the next PR.